### PR TITLE
[Merged by Bors] - feat: restriction of scalars for (pre)sheaves of modules

### PR DESCRIPTION
**Pull request by joelriou**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (callesonne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:32]:
```suggestion
    { toFun := M'.map f
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (callesonne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:33]:
```suggestion
      map_add' := map_add _
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (callesonne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:49]:
```suggestion
induced by a morphism of presheaves of rings `R ⟶ R'`. -/
```

File: Mathlib/Algebra/Category/ModuleCat/Sheaf/ChangeOfRings.lean
Comment (callesonne) [inline Mathlib/Algebra/Category/ModuleCat/Sheaf/ChangeOfRings.lean:29]:
```suggestion
induced by a morphism of sheaves of rings `R ⟶ R'`. -/
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (callesonne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:38]:
```suggestion
```
This line doesn't seem to be strictly necessary, but maybe this is not a good idea because too much burden is placed on `rfl` and I have no idea how efficient that is.

**Review by callesonne — COMMENTED** [review]
Apart from my comments, it all looks good to me. Can I approve it after these changes? Or should I wait for someone more experienced to also review it?

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:28]: It could be, but the file is very short, and doing so would require making also `M'` a global variable, otherwise the order of the parameters would have to be changed.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:38]: I have used the syntax `dsimp at eq ⊢` which also reduces the number of lines of code by `1`!

**Comment by joelriou** [comment]
Thanks @callesonne for your suggestions!

**Comment by joelriou** [comment]
> Can I approve it after these changes? Or should I wait for someone more experienced to also review it?

In general, you may very well approve PRs: it shall be a useful information for mathlib reviewers and maintainers (the later are the only ones that are allowed to merge PRs).

**Review by callesonne — APPROVED** [review]
This looks good to me now!

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (erdOne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:49]: Should this get `@[simps]`?

File: Mathlib/Algebra/Category/ModuleCat/Sheaf/ChangeOfRings.lean
Comment (erdOne) [inline Mathlib/Algebra/Category/ModuleCat/Sheaf/ChangeOfRings.lean:30]: Also this.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:50]: Thanks for the suggestion.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean
Comment (callesonne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:36]:
```suggestion
        erw [(M'.map_smul f (α.app _ r) x)]
```
I haven't tested this locally yet, but maybe it's a good idea to have this as a `erw` based on this [thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60erw.60.20vs.20.60apply.60).

Comment (callesonne) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/ChangeOfRings.lean:36]: Okay based on the continuation of the discussion, feel free to ignore this

**Comment by erdOne** [comment]
Thanks! (and also thanks Calle for reviewing!)
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
